Enhancements to the mustache script engine

martijnvg opened the pull request:
Enhancements:
* Added an option at compile time to decide variable values are json encoded or not encoded at all.
* Added support for variable placeholders to lookup values from specific slots in arrays/lists.

uboness commented on modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngineService.java:
```diff
@@ -85,8 +88,21 @@ public MustacheScriptEngineService(Settings settings) {
-        /** Factory to generate Mustache objects from. */
-        return (new JsonEscapingMustacheFactory()).compile(new FastStringReader(template), "query-template");
+        boolean escapeJson = true;
+        String escapeJsonStr = params.get(ESCAPE_JSON_PARAM);
```
I don't think that the scripting/template service/engine should be json oriented by default. I think it should work with potentially any content type (json, yaml, plain text, or whatever). I wouldn't have a parameter called `escape_son`, but instead maybe have a `content_type` parameter? and expect the values of this to be actual content types (i.e. `text/plain`, `application/json`, `application/yaml`, etc...)? By default we should simply assume `text/plain` and `utf8`. Then it'll be the responsibility of the client of this service to provide the context (e.g. the search templates will add `content_type=application/json`).

martijnvg replied:
The main reason I chose the `escape_json` and default to true approach is for isolating this change just to the `lang-mustache` module. Changing this to be based on content type and defaulting to be plain text, would mean that this change gets larger, while the default for ES is escaping json. Also currently there is only json escaping or no escaping. 

I'm not against this approach, so I'll try this out and see how much the impact is.

martijnvg commented:
@uboness I've updated the PR with your suggestion. There is only one issue with making plain text the default and that is for file based scripts and this is because compile arguments are used as cache key for file based scripts and the logic for loading file based scripts is generic for all script types, for this reason I had to add a hack that if the script type is mustache then the `content_type` is set to `application/json`. First of all this check is in the wrong place, but also this would be bad for plugins that use file based templating in non json content. In all other places it is easy to use decide where to add content type to the compile arguments, basically on all places where `Template` is being used, so I added a helper method for that on ScriptService.

The temporary if statement I added to ScriptService should be removed. Instead we should add an optional syntax that can be added at the beginning of each mustache template that controls the content type. I also think that in the case this optional syntax is absent we should default to `application/json` in the mustache plugin otherwise this would be a breaking change. Then all the changes I added in the last 2 commits can be reverted.

uboness commented:
wondering... shouldn't the content type be based on the file extension?

martijnvg commented:
@uboness That idea crossed my mind too, but this isn't something we do now. Right now, json encoding applies for all files. In the case that someone is relying on json encoding of the parameters in file based mustache script then this would mean that during upgrade files would have to be renamed too. Also it is kind of weird too that then mustache files would need to have the .json extension.

uboness commented:
I c... :/... this is really broken IMO... template & scripts are really two different things yet we treat them the same... not sure what's the "right" thing to do here in the current context. IMO the right thing to do is to really separate the two... but that's quite a big change for this PR. @s1monw @rmuir wdyt?

martijnvg commented:
yes, templates and scripts are two different animals each with their use cases. Mustache templates shouldn't be ran in a script query or aggregation and scripts shouldn't be use to template search requests. 

Maybe we should try to get this in? Because this change does clean things up too. We can do what I did initially, but just replace the `escape_json` with `content_type` and then default to json. 

In the followup PR we should make `Template` not extend from `Script` to begin with and separate the template specific infrastructure from the script service. There are overlapping things like location of templates and script, which can both be inline, file or a document in an index.

martijnvg commented:
I've updated the PR to change the default content type used for escaping back to `application/json` to make this change less invasive. In the future we should look into splitting of templates from scripts. 

@uboness Would you like to take a look?

uboness commented:
LGTM
